src/ring_theory/tensor_product.lean

```diff
@@ -1,36 +1,321 @@
+@[simps] def uncurry' : (M →ₗ[A] (N →ₗ[R] P)) →ₗ[A] ((M ⊗[R] N) →ₗ[A] P) :=
+{ to_fun := lift',
+  map_add' := λ f g, ext $ λ x y, by simp only [lift'_tmul, add_apply],
+  map_smul' := λ c f, ext $ λ x y, by simp only [lift'_tmul, smul_apply'] }
+
+/-- Heterobasic version of `tensor_product.lcurry`:
+
+Given a linear map `M ⊗[R] N →[A] P`, compose it with the canonical
+bilinear map `M →[A] N →[R] M ⊗[R] N` to form a bilinear map `M →[A] N →[R] P`. -/
+@[simps] def lcurry' : ((M ⊗[R] N) →ₗ[A] P) →ₗ[A] (M →ₗ[A] (N →ₗ[R] P)) :=
+{ to_fun := curry',
+  map_add' := λ f g, rfl,
+  map_smul' := λ c f, rfl }
+
+/-- Heterobasic version of `tensor_product.lift_equiv`:
+
+A linear equivalence constructing a linear map `M ⊗[R] N →[A] P` given a
+bilinear map `M →[A] N →[R] P` with the property that its composition with the
+canonical bilinear map `M →[A] N →[R] M ⊗[R] N` is the given bilinear map `M →[A] N →[R] P`. -/
+def lift_equiv' : (M →ₗ[A] (N →ₗ[R] P)) ≃ₗ[A] ((M ⊗[R] N) →ₗ[A] P) :=
+linear_equiv.of_linear (uncurry' R A M N P) (lcurry' R A M N P)
+  (linear_map.ext $ λ f, ext $ λ x y, lift'_tmul _ x y)
+  (linear_map.ext $ λ f, linear_map.ext $ λ x, linear_map.ext $ λ y, lift'_tmul f x y)
+
+variables {R A M N P}
+lemma curry'_inj : function.injective (@curry' R A M N P _ _ _ _ _ _ _ _ _ _ _ _ _) :=
+(lift_equiv' R A M N P).to_equiv.symm.injective
+
+variables (R A M N P)
+/-- Heterobasic version of `tensor_product.mk`:
+
+The canonical bilinear map `M →[A] N →[R] M ⊗[R] N`. -/
+@[simps] def mk' : M →ₗ[A] N →ₗ[R] M ⊗[R] N :=
+{ map_smul' := λ c x, rfl,
+  .. mk R M N }
+
+end comm_semiring
+
+section comm_semiring
+variables [comm_semiring R] [comm_semiring A] [algebra R A]
+variables [add_comm_monoid M] [semimodule A M]
+variables [add_comm_monoid N] [semimodule R N] [semimodule A N] [is_scalar_tower R A N]
+variables [add_comm_monoid P] [semimodule R P]
+
+/-- Heterobasic version of `tensor_product.assoc`:
+
+Linear equivalence between `(M ⊗[A] N) ⊗[R] P` and `M ⊗[A] (N ⊗[R] P)`. -/
+def assoc : ((M ⊗[A] N) ⊗[R] P) ≃ₗ[A] (M ⊗[A] (N ⊗[R] P)) :=
+linear_equiv.of_linear
+  (lift' $ uncurry A _ _ _ $ comp (lcurry' R A _ _ _) $ mk A M (N ⊗[R] P))
+  (uncurry A _ _ _ $ comp (uncurry' R A _ _ _) $ by apply curry; exact (mk' R A _ _))
+  (by apply mk_compr₂_inj; ext; refl)
+  (by apply curry'_inj; ext; refl)
+
+end comm_semiring
+
+end algebra_tensor_module
+
+end tensor_product
+
+namespace linear_map
 open tensor_product
 
+/-!
+### The basechange of a linear map of `R`-modules to a linear map of `A`-modules
+-/
+
+section semiring
+
+variables {R A B M N : Type*} [comm_semiring R]
+variables [semiring A] [algebra R A] [semiring B] [algebra R B]
+variables [add_comm_monoid M] [semimodule R M] [add_comm_monoid N] [semimodule R N]
+variables (f g : M →ₗ[R] N)
+
+variables (A)
+
+/-- `base_change A f` for `f : M →ₗ[R] N` is the `A`-linear map `A ⊗[R] M →ₗ[A] A ⊗[R] N`. -/
+def base_change (f : M →ₗ[R] N) : A ⊗[R] M →ₗ[A] A ⊗[R] N :=
```

Comment:
Should this have type `(M →ₗ[R] N) →ₗ[R] (A ⊗[R] M →ₗ[A] A ⊗[R] N)`?

Reply:
Nope, because then you can no longer use dot.notation
But having such a bundled version is useful. So if it hasn't been added, then we should.

Reply:
Could you add the linear map (as a new definition)?

Reply:
Done.